aten/src/ATen/native/AdaptiveAveragePooling3d.cpp

```diff
@@ -0,0 +1,314 @@
+        });
+  } else {
+    output.resize_({input.size(-5), sizeD, osizeT, osizeH, osizeW});
+    int64_t b;
+#pragma omp parallel for private(b)
+    for (b = 0; b < input.size(0); b++) {
+      AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+          input.scalar_type(), "adaptive_avg_pool3d_cpu", [&] {
+            auto input_data = input.data<scalar_t>();
+            auto output_data = output.data<scalar_t>();
+            adaptive_avg_pool3d_out_frame<scalar_t>(
+                input_data + b * input.stride(0),
+                output_data + b * sizeD * osizeT * osizeH * osizeW,
+                sizeD,
+                isizeT,
+                isizeH,
+                isizeW,
+                osizeT,
+                osizeH,
+                osizeW,
+                istrideD,
+                istrideT,
+                istrideH,
+                istrideW);
+          });
+    }
+  }
+}
+
+template <typename scalar_t>
+static void adaptive_avg_pool3d_backward_out_frame(
+    scalar_t* gradInput_p,
+    scalar_t* gradOutput_p,
+    int64_t sizeD,
+    int64_t isizeT,
+    int64_t isizeH,
+    int64_t isizeW,
+    int64_t osizeT,
+    int64_t osizeH,
+    int64_t osizeW) {
+  int64_t d;
+#pragma omp parallel for private(d)
+  for (d = 0; d < sizeD; d++) {
+    scalar_t* gradInput_p_d = gradInput_p + d * isizeT * isizeW * isizeH;
+    scalar_t* gradOutput_p_d = gradOutput_p + d * osizeT * osizeW * osizeH;
+
+    /* calculate average */
+    int64_t ot, oh, ow;
+    for (ot = 0; ot < osizeT; ot++) {
+      int istartT = start_index(ot, osizeT, isizeT);
+      int iendT = end_index(ot, osizeT, isizeT);
+      int kT = iendT - istartT;
+
+      for (oh = 0; oh < osizeH; oh++) {
+        int istartH = start_index(oh, osizeH, isizeH);
+        int iendH = end_index(oh, osizeH, isizeH);
+        int kH = iendH - istartH;
+
+        for (ow = 0; ow < osizeW; ow++) {
+          int istartW = start_index(ow, osizeW, isizeW);
+          int iendW = end_index(ow, osizeW, isizeW);
+          int kW = iendW - istartW;
+
+          scalar_t grad_delta =
+              gradOutput_p_d[ot * osizeH * osizeW + oh * osizeW + ow] / kT /
+              kH / kW;
+
+          int it, ih, iw;
+          for (it = istartT; it < iendT; it++) {
+            for (ih = istartH; ih < iendH; ih++) {
+              for (iw = istartW; iw < iendW; iw++) {
+                /* update gradient */
+                gradInput_p_d[it * isizeH * isizeW + ih * isizeW + iw] +=
+                    grad_delta;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+Tensor& adaptive_avg_pool3d_backward_out_cpu_template(
+    Tensor& gradInput,
+    const Tensor& gradOutput_,
+    const Tensor& input) {
+  /* sizes */
+  int sizeD = input.size(-4);
+  int isizeT = input.size(-3);
+  int isizeH = input.size(-2);
+  int isizeW = input.size(-1);
+  int osizeT = gradOutput_.size(-3);
+  int osizeH = gradOutput_.size(-2);
+  int osizeW = gradOutput_.size(-1);
+
```

Comment:
The old code had `int64_t` here.  `AdaptiveAveragePooling.cpp` seems to use `int`, too, but in general I'd hesitate to change integer sizes during porting.